src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingProcessor.cs

```diff
@@ -0,0 +1,418 @@
+                while (!innerCts.IsCancellationRequested)
+                {
+                    await Task.Delay(5000);
+                    await AcquirePartitions(innerCts.Token);
+                    await RunAndRenew(innerCts);
+                }
+            }
+        }
+
+        private Task ContinuallyRegisterProcessor()
+        {
+            Task.Run(() => RegisterProcessorLoop());
+            return Task.CompletedTask;
+        }
+
+        private Task CheckProcessorCount(CancellationTokenSource cts)
+        {
+            Task.Run(async () =>
+            {
+                while (!cts.IsCancellationRequested)
+                {
+                    var activePods = ListBlobs(_containerClient, "activepods");
+                    var activeProcessors = await GetActiveProcessors(activePods, _containerClient);
+
+                    Console.WriteLine($"CheckProcessorCount - active processors {activeProcessors}");
+
+                    if (activeProcessors != _totalProcessorsRunning)
+                    {
+                        Console.WriteLine($"Detected that number of processors equals {activeProcessors} but this processor is running with {_totalProcessorsRunning} known processors");
+                        Console.WriteLine("Restarting and recomputing");
+                        cts.Cancel();
+                    }
+
+                    await Task.Delay(30000);
+                }
+            });
+
+            return Task.CompletedTask;
+        }
+
+        private Task RenewOwnership(CancellationToken ct)
+        {
+            Task.Run(async () =>
+            {
+                while (!ct.IsCancellationRequested)
+                {
+                    foreach (var partition in _ownedPartitions)
+                    {
+                        Console.WriteLine($"Renewing ownership for partition {partition}");
+
+                        // get the blob and ETag
+                        BlobClient ownershipClient = _containerClient.GetBlobClient($"ownedpartitions/{partition}");
+                        var properties = await ownershipClient.GetPropertiesAsync();
+                        var eTag = properties.Value.ETag;
+
+                        BlobUploadOptions ownershipClientOptions = new BlobUploadOptions()
+                        {
+                            Metadata = new Dictionary<string, string>()
+                            {
+                                { "owner", _processorId },
+                                { "owneduntil", DateTime.UtcNow.AddMinutes(1).ToString() },
+                            },
+                            Conditions = new BlobRequestConditions()
+                            {
+                                IfMatch = eTag,
+                            },
+                        };
+
+                        var ownershipResult = await ownershipClient.UploadAsync(BinaryData.FromString(string.Empty), ownershipClientOptions);
```

Comment:
Do we have to upload the blob again to refresh ownership?  It looks like the lease blob might be better for this, https://learn.microsoft.com/en-us/rest/api/storageservices/lease-blob

Reply:
agree, leases would be better. can switch to those

Reply:
switched to blob leases for ownership, thanks for the suggestion